### PR TITLE
CA-327382: reap child processes

### PR DESCRIPTION
**Pull request by MarkSymsCtx**
This caused XSI-474 and other reported customer issues.

Signed-off-by: Mark Syms <mark.syms@citrix.com>

### CODE REVIEW

File: drivers/cleanup.py [hunk 1]
```diff
@@ -147,22 +147,35 @@ def runAbortable(func, ret, ns, abortTest, pollInterval, timeOut):
...
+                        raise AbortException("Aborting due to signal")
+                    if timeOut and time.time() - startTime > timeOut:
+                        os.killpg(pid, signal.SIGKILL)
+                        resultFlag.clearAll()
+                        raise util.SMException("Timed out")
+                    time.sleep(pollInterval)
+            finally:
+                wait_pid = 0
+                rc = -1
+                count = 0
```
Comment (BenSimsCitrix) [inline drivers/cleanup.py]: Could we simplify the code by having a resultFlag.clearAll() in the finally?

Comment (MarkSymsCtx) [reply drivers/cleanup.py]: Actually, the whole resultflag piece can be removed, but that's a bigger change for another day. We can just use the exit code from the process by moving the `waitpid(pid, WNOHANG)` inside the loop, basically poll the subprocess until it completes and the read the exit status.

File: drivers/cleanup.py [hunk 2]
```diff
@@ -147,22 +147,35 @@ def runAbortable(func, ret, ns, abortTest, pollInterval, timeOut):
...
+                        resultFlag.clear("success")
+                        return
+                    if resultFlag.test("failure"):
+                        resultFlag.clear("failure")
+                        raise util.SMException("Child process exited with error")
+                    if abortTest() or abortSignaled:
+                        os.killpg(pid, signal.SIGKILL)
+                        raise AbortException("Aborting due to signal")
+                    if timeOut and time.time() - startTime > timeOut:
+                        os.killpg(pid, signal.SIGKILL)
```
Comment (BenSimsCitrix) [inline drivers/cleanup.py]: perhaps we could tidyup by having a resultFlag.clearall() in the finally.

File: drivers/cleanup.py [hunk 3]
```diff
@@ -147,22 +147,35 @@ def runAbortable(func, ret, ns, abortTest, pollInterval, timeOut):
...
+                    time.sleep(pollInterval)
+            finally:
+                wait_pid = 0
+                rc = -1
+                count = 0
+                while wait_pid == 0 and count < 10:
+                    wait_pid, rc = os.waitpid(pid, os.WNOHANG)
+                    if wait_pid == 0:
+                        time.sleep(2)
+                        count += 1
```
Comment (BenSimsCitrix) [inline drivers/cleanup.py]: docs suggest that an OSError can be raised, do we want to catch it or bomb out.

File: drivers/cleanup.py (hunk 2 above)
Comment (TimSmithCtx) [reply drivers/cleanup.py]: > perhaps we could tidyup by having a resultFlag.clearall() in the finally.

Do we still care at that point? If we entered the finally: as the result of an exception, the exception will be re-thrown when the finally completes.

**Pull request merged**